kyuubi-server/src/main/scala/org/apache/kyuubi/client/KyuubiSyncThriftClient.scala

```diff
@@ -55,6 +57,47 @@ class KyuubiSyncThriftClient private (protocol: TProtocol)
     } finally lock.unlock()
   }
 
+  private def withRetryingRpc[T](block: => T, rpcName: String): T = {
+    var retryCount = 0
+
+    def shouldRetry(e: Throwable): Boolean = {
+      e.isInstanceOf[TException] && retryCount < maxRetries && !isClosed() && isConnectionValid()
+    }
+
+    def isConnectionValid(): Boolean = {
+      var isValid = false
+      val tGetInfoReq = new TGetInfoReq()
+      tGetInfoReq.setSessionHandle(_remoteSessionHandle)
+      tGetInfoReq.setInfoType(TGetInfoType.CLI_DBMS_VER)
+      try {
+        withLockAcquired(GetInfo(tGetInfoReq).getInfoValue.getStringValue)
+        isValid = true
+      } catch {
+        case e: Throwable =>
+          error(s"The internal connection is not valid, mark the thrift client as closed", e)
+          setClosed()
+      }
+      isValid
+    }
+
+    try {
+      withLockAcquired(block)
```

Comment:
It seems the lock should be high level. It shoud be atomical even if we retry one operation with several rpcs. That said  the lock is working between operation and operation, not rpc and rpc.

Reply:
thanks